src/components/EngineStream.tsx

```diff
@@ -142,51 +213,65 @@ export const EngineStream = (props: {
   useEffect(() => {
     engineCommandManager.addEventListener(
       EngineCommandManagerEvents.SceneReady,
-      play
+      executeKcl
     )
     return () => {
       engineCommandManager.removeEventListener(
         EngineCommandManagerEvents.SceneReady,
-        play
+        executeKcl
       )
     }
   }, [firstPlay])
 
   useEffect(() => {
-    engineCommandManager.addEventListener(
-      EngineCommandManagerEvents.SceneReady,
-      play
-    )
+    // We do a back-off restart, using a fibonacci sequence, since it
+    // has a nice retry time curve (somewhat quick then exponential)
+    const attemptRestartIfNecessary = () => {
+      if (isRestartRequestStarting) return
+      setIsRestartRequestStarting(true)
+      setTimeout(() => {
+        engineStreamState.context.videoRef.current?.pause()
+        engineCommandManager.tearDown()
+        startOrReconfigureEngine()
+        setFirstPlay(false)
+        setIsRestartRequestStarting(false)
+      }, attemptTimes[0] + attemptTimes[1])
+      setAttemptTimes([attemptTimes[1], attemptTimes[0] + attemptTimes[1]])
+    }
 
-    engineStreamActor.send({
-      type: EngineStreamTransition.SetPool,
-      data: { pool: props.pool },
-    })
-    engineStreamActor.send({
-      type: EngineStreamTransition.SetAuthToken,
-      data: { authToken: props.authToken },
-    })
+    // Poll that we're connected. If not, send a reset signal.
+    // Do not restart if we're in idle mode.
+    const connectionCheckIntervalId = setInterval(() => {
+      // SKIP DURING TESTS BECAUSE IT WILL MESS WITH REUSING THE
+      // ELECTRON INSTANCE.
+      if (TEST) {
+        return
+      }
 
-    return () => {
-      engineCommandManager.tearDown()
-    }
-  }, [])
+      // Don't try try to restart if we're already connected!
+      const hasEngineConnectionInst = engineCommandManager.engineConnection
+      const isDisconnected =
+        engineCommandManager.engineConnection?.state.type ===
+        EngineConnectionStateType.Disconnected
+      const inIdleMode = engineStreamState.value === EngineStreamState.Paused
+      if ((hasEngineConnectionInst && !isDisconnected) || inIdleMode) return
 
-  // In the past we'd try to play immediately, but the proper thing is to way
-  // for the 'canplay' event to tell us data is ready.
-  useEffect(() => {
-    const videoRef = engineStreamState.context.videoRef.current
-    if (!videoRef) {
-      return
-    }
-    const play = () => {
-      videoRef.play().catch(console.error)
-    }
-    videoRef.addEventListener('canplay', play)
+      attemptRestartIfNecessary()
+    }, 1000)
```

Comment:
Might want to put this 1000 in a const as well to make it more descriptive of what it is?